internal/app/secretless/proxy.go

```diff
@@ -75,50 +64,87 @@ func (p *Proxy) Listen(listenerConfig config.Listener, wg sync.WaitGroup) plugin
 	p.EventNotifier.CreateListener(listener)
 
 	go func() {
-		defer wg.Done()
+		defer listener.Shutdown()
 		listener.Listen()
 	}()
 
 	return listener
 }
 
-// ReloadListeners will loop through the listeners and shut them down
-// As each listener is shut down the WaitGroup is decremented, and once the
-// counter is zero the Proxy.Run loop will complete and restart, reloading all
-// of the listeners.
+// ReloadListeners sends RESTART msg to runEventChan
 func (p *Proxy) ReloadListeners() error {
-	if p.Listeners == nil || len(p.Listeners) == 0 {
-		log.Println("WARN: No listeners to reload!")
-		return nil
-	}
+	p.runEventChan <- RESTART
+
+	// TODO: Return any errors we get during reload
+	return nil
+}
+
+// Shutdown sends SHUTDOWN msg to runEventChan
+func (p *Proxy) Shutdown() {
+	p.runEventChan <- SHUTDOWN
+	p.cleanUpListeners()
+}
+
+// Loops through the listeners and shuts them down concurrently
+func (p *Proxy) cleanUpListeners() {
+	// because cleanUpListeners can be called from different goroutines
+	defer p.cleanupMutex.Unlock()
+	p.cleanupMutex.Lock()
+
+	var wg sync.WaitGroup
 
 	for _, listener := range p.Listeners {
+		_listener := listener
+
 		log.Printf("Shutting down '%v' listener...", listener.GetName())
-		listener.Shutdown()
-		p.ListenerWaitGroup.Done()
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			_listener.Shutdown()
+		}()
 	}
 
-	// TODO: Return any errors we get during reload
-	return nil
+	wg.Wait()
 }
 
 // Run is the main entrypoint to the secretless program.
+// the for-select loop allows for queueing of RESTARTS and only 1 SHUTDOWN
 func (p *Proxy) Run() {
-	p.ListenerWaitGroup = sync.WaitGroup{}
-	// We loop until we get an exit signal (in which case we exit program)
-	for {
-		// TODO: Delegate logic of this `if` check to connection managers
-		if len(p.Config.Listeners) < 1 {
-			log.Fatalln("ERROR! No listeners specified in config!")
-		}
+	p.runEventChan = make(chan int, 1)
+	p.cleanupMutex = sync.Mutex{}
 
-		p.Listeners = make([]plugin_v1.Listener, 0)
-		log.Println("Starting all listeners and handlers...")
-		p.ListenerWaitGroup.Add(len(p.Config.Listeners))
-		for _, config := range p.Config.Listeners {
-			listener := p.Listen(config, p.ListenerWaitGroup)
-			p.Listeners = append(p.Listeners, listener)
+	go func() {
+		p.runEventChan <- START
+	}()
+
+	// when runEventChan receives message...
+	// RESTART: runs cleanUpListeners and reloads all listeners
+	// SHUTDOWN: for-select turns to infinite non-busy loop
+	// default: panic
+	for {
+		select {
+		case msg := <-p.runEventChan:
```

Comment:
I [believe](https://gobyexample.com/non-blocking-channel-operations) that this type of loop is a non-blocking one which means that this loop will eat up cycles instead of being an async triggered event so we should probably swap it to a blocking type.

Reply:
@sgnn7 `for-select` is blocking because `select` is blocking.

https://tour.golang.org/concurrency/5
https://stackoverflow.com/questions/48769062/whats-difference-between-select-and-for-in-golang